File: web/src/lib/zones.ts
import { CONFIG } from "./config";
import { volumeAtPrice } from "./volumeProfile";
import type { Candle, VolumeProfile, Zone } from "./types";

const clamp01 = (v: number) => Math.min(Math.max(v, 0), 1);
const mean = (a: number[]) => a.reduce((x, y) => x + y, 0) / (a.length || 1);

interface Candidate {
  price: number;
  time: number | null; // null for volume-node seeds
  source: string; // "pivot" | POC | VAH | VAL | HVN
}

function swingPivots(candles: Candle[], k: number) {
  const highs: { time: number; price: number }[] = [];
  const lows: { time: number; price: number }[] = [];
  for (let i = k; i < candles.length - k; i++) {
    let isHigh = true;
    let isLow = true;
    for (let j = i - k; j <= i + k; j++) {
      if (candles[j].high > candles[i].high) isHigh = false;
      if (candles[j].low < candles[i].low) isLow = false;
    }
    if (isHigh) highs.push({ time: candles[i].time, price: candles[i].high });
    if (isLow) lows.push({ time: candles[i].time, price: candles[i].low });
  }
  return { highs, lows };
}

function rejectionAt(
  candles: Candle[],
  lo: number,
  hi: number,
  atr: number,
): number {
  const rej: number[] = [];
  for (const c of candles) {
    if (c.low <= hi && c.high >= lo) {
      const bodyHi = Math.max(c.open, c.close);
      const bodyLo = Math.min(c.open, c.close);
      rej.push(Math.max(c.high - bodyHi, bodyLo - c.low));
    }
  }
  if (!rej.length || atr <= 0) return 0;
  return clamp01(mean(rej) / atr);
}

export function detectZones(
  candles: Candle[],
  profile: VolumeProfile,
  atr: number,
  price: number,
): Zone[] {
  const { highs, lows } = swingPivots(candles, CONFIG.pivotLookback);
  const cands: Candidate[] = [];
  for (const h of highs) cands.push({ price: h.price, time: h.time, source: "pivot" });
  for (const l of lows) cands.push({ price: l.price, time: l.time, source: "pivot" });
  cands.push({ price: profile.poc, time: null, source: "POC" });
  cands.push({ price: profile.vah, time: null, source: "VAH" });
  cands.push({ price: profile.val, time: null, source: "VAL" });
  for (const p of profile.hvns) cands.push({ price: p, time: null, source: "HVN" });

  cands.sort((a, b) => a.price - b.price);
  const tol = CONFIG.clusterAtrMult * atr;

  // Agglomerate candidates within `tol` of the running cluster mean.
  const clusters: Candidate[][] = [];
  for (const c of cands) {
    const last = clusters[clusters.length - 1];
    if (last && Math.abs(c.price - mean(last.map((x) => x.price))) <= tol) {
      last.push(c);
    } else {
      clusters.push([c]);
    }
  }

  const t0 = candles[0].time;
  const span = candles[candles.length - 1].time - t0 || 1;
  const w = CONFIG.weights;
  const zones: Zone[] = [];

  for (const cl of clusters) {
    const center = mean(cl.map((x) => x.price));
    const half = CONFIG.zoneWidthAtr * atr;
    const lo = center - half;
    const hi = center + half;
    const kind = center >= price ? "resistance" : "support";

    const tags = [...new Set(cl.filter((x) => x.source !== "pivot").map((x) => x.source))].sort();
    const touches = cl.filter((x) => x.source === "pivot").length;

    const volS = volumeAtPrice(profile, center);
    const touchS = Math.min(touches / 4, 1);
    const rejS = rejectionAt(candles, lo, hi, atr);
    const confS = Math.min(tags.length / 3, 1);
    const times = cl.filter((x) => x.time !== null).map((x) => x.time as number);
    const recencyS = times.length
      ? clamp01((Math.max(...times) - t0) / span)
      : 0.5;

    const strength =
      100 *
      (w.volume * volS +
        w.touches * touchS +
        w.rejection * rejS +
        w.confluence * confS +
        w.recency * recencyS);

    zones.push({
      price: center,
      lo,
      hi,
      kind,
      strength: Math.round(strength * 10) / 10,
      touches,
      tags,
      components: {
        volume: Math.round(volS * 100) / 100,
        touches: Math.round(touchS * 100) / 100,
        rejection: Math.round(rejS * 100) / 100,
        confluence: Math.round(confS * 100) / 100,
        recency: Math.round(recencyS * 100) / 100,
      },
    });
  }

  zones.sort((a, b) => b.strength - a.strength);
  return zones;
}

File: zones.py
"""Support/Resistance zone detection and strength scoring.

Pipeline:
  1. Find swing pivots (fractals) -> raw candidate levels.
  2. Cluster nearby pivots + volume nodes into zones.
  3. Score each zone 0-100 on volume, touches, rejection, confluence, recency.
"""
from __future__ import annotations
from dataclasses import dataclass, field
import numpy as np
import pandas as pd

import config
from volume_profile import VolumeProfile, volume_at_price


@dataclass
class Zone:
    price: float                 # zone center
    lo: float                    # band low
    hi: float                    # band high
    kind: str                    # "support" | "resistance"
    strength: float = 0.0        # composite 0-100
    touches: int = 0
    components: dict = field(default_factory=dict)
    tags: list[str] = field(default_factory=list)  # POC / VAH / HVN ...

    def contains(self, price: float) -> bool:
        return self.lo <= price <= self.hi


def _swing_pivots(df: pd.DataFrame, k: int):
    """Fractal swing highs and lows with k bars of confirmation each side."""
    highs, lows = [], []
    h, l = df["high"].values, df["low"].values
    t = df.index
    for i in range(k, len(df) - k):
        win_h = h[i - k:i + k + 1]
        win_l = l[i - k:i + k + 1]
        if h[i] == win_h.max() and (win_h.argmax() == k):
            highs.append((t[i], h[i]))
        if l[i] == win_l.min() and (win_l.argmin() == k):
            lows.append((t[i], l[i]))
    return highs, lows


def _rejection_at(df: pd.DataFrame, lo: float, hi: float, atr: float) -> float:
    """Mean ATR-normalized wick rejection for candles that tested the band."""
    rej = []
    for _, r in df.iterrows():
        if r["low"] <= hi and r["high"] >= lo:  # candle overlapped the zone
            body_hi = max(r["open"], r["close"])
            body_lo = min(r["open"], r["close"])
            upper_wick = r["high"] - body_hi
            lower_wick = body_lo - r["low"]
            rej.append(max(upper_wick, lower_wick))
    if not rej or atr <= 0:
        return 0.0
    return float(np.clip(np.mean(rej) / atr, 0, 1))


def detect_zones(df: pd.DataFrame, profile: VolumeProfile, atr: float,
                 price: float) -> list[Zone]:
    highs, lows = _swing_pivots(df, config.PIVOT_LOOKBACK)

    # Seed candidate levels: pivots + key volume-profile prices.
    candidates = []  # (price, time_or_None, source)
    for t, p in highs:
        candidates.append((p, t, "pivot"))
    for t, p in lows:
        candidates.append((p, t, "pivot"))
    for p, tag in [(profile.poc, "POC"), (profile.vah, "VAH"),
                   (profile.val, "VAL")]:
        candidates.append((p, None, tag))
    for p in profile.hvns:
        candidates.append((p, None, "HVN"))

    candidates.sort(key=lambda x: x[0])
    tol = config.CLUSTER_ATR_MULT * atr

    # Agglomerate candidates that sit within `tol` of each other.
    clusters: list[list] = []
    for c in candidates:
        if clusters and abs(c[0] - np.mean([x[0] for x in clusters[-1]])) <= tol:
            clusters[-1].append(c)
        else:
            clusters.append([c])

    last_time = df.index[-1]
    span = (last_time - df.index[0]).total_seconds() or 1.0
    zones: list[Zone] = []

    for cl in clusters:
        prices = [x[0] for x in cl]
        center = float(np.mean(prices))
        half = config.ZONE_WIDTH_ATR * atr
        lo, hi = center - half, center + half
        kind = "resistance" if center >= price else "support"

        tags = sorted({x[2] for x in cl if x[2] not in ("pivot",)})
        touches = sum(1 for x in cl if x[2] == "pivot")

        # --- component scores (each 0-1) ---
        vol_s = volume_at_price(profile, center)
        touch_s = min(touches / 4.0, 1.0)
        rej_s = _rejection_at(df, lo, hi, atr)
        conf_s = min(len(tags) / 3.0, 1.0)
        # recency: most recent pivot time in the cluster, time-decayed
        times = [x[1] for x in cl if x[1] is not None]
        if times:
            newest = max(times)
            recency_s = float((newest - df.index[0]).total_seconds() / span)
        else:
            recency_s = 0.5
        recency_s = np.clip(recency_s, 0, 1)

        w = config.WEIGHTS
        strength = 100 * (
            w["volume"] * vol_s +
            w["touches"] * touch_s +
            w["rejection"] * rej_s +
            w["confluence"] * conf_s +
            w["recency"] * recency_s
        )

        zones.append(Zone(
            price=round(center, 6), lo=round(lo, 6), hi=round(hi, 6),
            kind=kind, strength=round(strength, 1), touches=touches,
            tags=tags,
            components={
                "volume": round(vol_s, 2), "touches": round(touch_s, 2),
                "rejection": round(rej_s, 2), "confluence": round(conf_s, 2),
                "recency": round(recency_s, 2),
            },
        ))

    zones.sort(key=lambda z: z.strength, reverse=True)
    return zones
